File: algorithms/2_medium/C/02211_count_collisions_on_a_road/const_space.cpp
class Solution {
public:
    int countCollisions(string directions) {
        int count{0};
        int curr_r{0};
        int curr_s{0};

        for (const char& ch : directions) {
            switch (ch) {
                case 'L':
                    if (curr_r > 0) {
                        count += curr_r + 1;
                        curr_r = 0;
                        curr_s = 1;
                    } else if (curr_s != 0) {
                            count += 1;
                    }
                    break;
                case 'R':
                    curr_r += 1;
                    break;
                case 'S':
                    count += curr_r;
                    curr_r = 0;
                    curr_s = 1;
                    break;
                default:
                    break;
            }
        }

        return count;
    }
};

File: algorithms/2_medium/C/02211_count_collisions_on_a_road/const_space.rs
impl Solution {
    pub fn count_collisions(directions: String) -> i32 {
        let mut count = 0_i32;
        let mut curr_r = 0_i32;
        let mut curr_s = 0_i32;

        for byte in directions.as_bytes().iter() {
            match *byte {
                b'L' => {
                    if curr_r > 0 {
                        count += curr_r + 1;
                        curr_r = 0;
                        curr_s = 1;
                    } else if curr_s != 0 {
                            count += 1;
                    }
                },
                b'R' => {
                    curr_r += 1;
                },
                b'S'=> {
                    count += curr_r;
                    curr_r = 0;
                    curr_s = 1;
                },
                _ => {},
            }
        }

        count
    }
}
